File: frontend/src/types/transaction.ts
import type { Account } from "./account"

export type TransactionType = "income" | "expense"

export interface TransactionForm {
  type: TransactionType
  amount: number | null
  category: string
  accountId: number | null
  // title: string
  note: string
  transactionDate: string
  slipImage: File | null
}

export interface Transaction {
  id: number
  // account: Account
  accountId: number
  type: TransactionType
  amount: number
  category: string
  bankName: string
  // title: string
  note: string
  transactionDate: string
  slipImage?: string | null
}

export interface TransactionFilter {
    accountId: number | null
    year?: number | null
    month?: number | null
    startDate?: string | null
    endDate?: string | null
    category?: string | null
}

export interface CreateTransactionResponse {
  message: string
  transaction: Transaction
}

export interface GetTransactionsResponse {
  transactions: Transaction[]
}

export interface DeleteTransactionResponse {
  message: string
}

File: backend/internal/dto/transaction.go
package dto

import (
	"backend/internal/models"
	
	"time"
)

type TransactionInput struct {
	Type            string
	Amount          float64
	Category        string
	AccountID       uint
	Title           string
	Note            string
	TransactionDate time.Time
}

type TransactionFilter struct {
	StartDate time.Time
	EndDate   time.Time // วันสิ้นสุดแบบ exclusive
	AccountID *uint
	Category  string
	Type      string
}

type TransactionResponse struct {
	ID              uint      `json:"id"`
	AccountID       uint      `json:"accountId"`
	Type            string    `json:"type"`
	Amount          float64   `json:"amount"`
	Category        string    `json:"category"`
	Title           string    `json:"title"`
	Note            string    `json:"note"`
	Image           string    `json:"image"`
	TransactionDate time.Time `json:"transactionDate"`
	CreatedAt       time.Time `json:"createdAt"`
}

func ToTransactionResponse(
	transaction models.Transaction,
) TransactionResponse {
	return TransactionResponse{
		ID:              transaction.ID,
		AccountID:       transaction.AccountID,
		Type:            transaction.Type,
		Amount:          transaction.Amount,
		Category:        transaction.Category,
		Title:           transaction.Title,
		Note:            transaction.Note,
		Image:           transaction.Image,
		TransactionDate: transaction.TransactionDate,
		CreatedAt:       transaction.CreatedAt,
	}
}

